content/blog/2019-10-02-codestitcher.md

```diff
@@ -0,0 +1,166 @@
++++
+title = "Codestitcher: inter-procedural basic block layout optimization"
+extra.bio = """
+Eashan Garg is an undergraduate studying CS and English.
+Gautam Mekkat is an undergraduate studying CS and ECE.
+"""
+extra.author = "Eashan Garg and Gautam Mekkat"
++++
+
+## Introduction
+
+So far, we've covered traditional compiler optimizations such as dead code elimination, copy propagation, and constant propagation. These techniques are centered around *elimination*--- reducing the number of instructions in a program. This is useful, but programs have dramatically increased in size over the years. Eliminating unnecessary code can only do so much. Meanwhile, the capacities of memory units like L1 caches and the TLB haven't grown to account for larger program sizes. One reason for this is that as cache size increases, latency increases. Because of this, we want to optimally place instructions in caches, making sure the most commonly run code lives in the cache. Code layout is critical here. In this post we will discuss this paper's approach to code layout.
+
+## Background: Optimizing Code Layout
+
+Before we get too far in, let's consider how modern compilers approach code layout, specifically function placement. Since 1990, [Pettis and Hansen's code reordering method](https://dl.acm.org/citation.cfm?id=93550) has been the de facto approach, used by LLVM, HHVM, and other compiler infrastructures. Recent function layout techniques, such as call-chain clustering, attempt to improve upon the Pettis-Hansen method by addressing some of its limitations.
+
+### Pettis-Hansen Method
+Informally, given a program, the Pettis-Hansen method (hereafter referred to as PH for brevity) works as follows.
+
+1. Generate an undirected call graph, where nodes are functions and edges represent calls from functions to other functions. More formally, a function A and a function B would have an edge between each other if A calls B or B calls A.
+2. Run a benchmark workload on this program, keeping track of what calls are made.
+3. Generate a weighted call graph, where edge weights are proportional to the frequency of calls between the two corresponding functions.
+4. Go through the edges in decreasing weight, joining the nodes connected by the edge. When nodes join, their edges combine, summing weights when necessary. Node names are also concatenated, denoting the corresponding function order.
+    - The node names concatenate to maximize the sum of the weights between consecutive nodes. This can involve reversing node names.
+6. After this process is complete, no edges will remain in the graph, and the order in which nodes are joined denote the function order.
+
+Let's go through a brief example. Suppose the weighted call graph is as follows.
+
+<img src="https://i.imgur.com/tvee3J0.png" width="400px">
```

Comment:
Hotlinking Imgur-hosted images doesn't seem to work for me. They get 403 errors. (Did it work for you?)
<img width="820" alt="Screen Shot 2019-10-03 at 3 15 57 PM" src="https://user-images.githubusercontent.com/188033/66156955-b8fcd580-e5f0-11e9-8a63-5af234d11551.png">

You can put images right here in the 6120 website. Just make you post a directory instead of a file, and call the body `index.md`. Then you can put images in the same directory.